[IMP] MRP: Multilevel BoM updates + setup flow (#12290)
odoo/documentation: content/applications/inventory_and_mrp/manufacturing/advanced_configuration/sub_assemblies.rst

Sublevel |BOMs| (subassemblies or semifinished products) streamline these production workflows, and
are beneficial when the subassembly is used across multiple finished products (meaning they'd appear
on multiple top-level |BOMs|). The more complicated a product is to build or procure, the more value
a multilevel |BOM| can provide. Because of this, component and subassembly replenishment planning is
vital to ensuring multilevel |BOMs| run smoothly.

Why replenishment planning is important
=======================================

:doc:`Replenishment <../../inventory/warehouses_storage/replenishment>` is critical for multilevel
BoMs to prevent bottlenecks, manage lead times, and optimize inventory. Without it, missing
components can halt production, delay orders, and increase costs. A well-planned replenishment

> strategy ensures
> 
> 
> 
> :doc:`just-in-time availability <../../inventory/warehouses_storage/replenishment/reordering_rules>`
> , :doc:`automates procurement <../../inventory/warehouses_storage/replenishment/reordering_rules>`,
> balances stock levels, and keeps supply chains efficient. This minimizes delays, reduces manual
> effort, and ensures smooth manufacturing.

Reviewer: Please fix this accidental gap in the RST
```suggestion
strategy ensures :doc:`just-in-time availability
<../../inventory/warehouses_storage/replenishment/reordering_rules>`, :doc:`automates procurement
<../../inventory/warehouses_storage/replenishment/reordering_rules>`, balances stock levels, and
keeps supply chains efficient. This minimizes delays, reduces manual effort, and ensures smooth
manufacturing.
```

Manufacturing orders (MO)s that come from a |BOM| require all components to be available before the
|MO| can be completed. The :guilabel:`Component Status` for that |MO| provides this information.
Learn how to :doc:`check an MO's component status <../basic_setup/bill_configuration>`

> .. _Create a Multilevel BoM:

Reviewer: Custom anchors must not contain spaces and are to follow [a convention](https://www.odoo.com/documentation/master/contributing/documentation/rst_guidelines.html#custom-anchors)
```suggestion
.. _manufactuing/advanced/create-multilevel-bom:
```

If starting from scratch, build the |BOMs| from the bottom up. Start with the lowest-level product
|BOMs|, then include those products as components in higher-level |BOMs|.

> For example, a printed circuit board (PCB) for a custom keyboard is composed of hundreds of

Reviewer 2: You say 'For example', which is the perfect opportunity for an example admonition block. I always love the analogy of cars since everyone has been in a car and understands how complex they are. So there are the sub-assemblies for: engines, seats, doors, etc. It's super-relatable and has TONS of assemblies, and I think people have been exposed to so many car commercials where we see the components and factories; it's very relatable. Not that the keyboard isn't, but the car is, I feel, the most universal example of complex, multi-level BOMs. Just a suggestion - no need to take it!

Reordering rules are recommended because they do not tie the manufactured product to a specific
sales order, allowing that manufactured product to fulfill a different sales order if the original
gets cancelled.

Making the product to order is not recommended because the manufactured product cannot be used to
fulfill another sales order. However, this could be helpful if that strict tracking is necessary for
the business.

Enter vendor and manufacturing lead times
-----------------------------------------

Vendor and manufacturing lead times will lead to better predictions for production and order
fulfillment. Set vendor lead times for components that are purchased. These can appear at any level
in a multilevel |BOM| except the final product. Set manufacturing lead times for products that are
built using a |BOM|. These can appear at any level in a multilevel |BOM| except the lowest level
(when individual components are procured).

.. tip::
   Learn more about :doc:`vendor lead times and manufacturing lead times
   <../../inventory/warehouses_storage/replenishment/lead_times>`.

Build out operations to handle the production flow
--------------------------------------------------

> This step is the most unique part of the multilevel |BOM| setup. Take the time to understand what
> this looks like for the business, then match the Odoo setup to that real-life setup. The list below
> are just some of the configuration pieces that could be involved in this step.

Reviewer 2: ```suggestion
First, determine the current manufacturing flow for the business, then match the corresponding Odoo setup. The list below
are just some of the configuration pieces that could be involved in this step.
```

  three-step manufacturing).
- **Work Centers**: Decide if any :doc:`work centers <using_work_centers>` need to be configured.
- **Master Production Schedules**: If manually planned manufacturing orders are needed (e.g., to
  handle seasonal demand), make a :doc:`master production schedule <../workflows/use_mps>` (MPS).

.. tip::
   Manufacturing operations is an art and a science, so configuring an established flow into Odoo is
   the recommended approach for this step. Read more about :doc:`manufacturing in Odoo
   <../../manufacturing>`

Configuration summary
---------------------

At the end of this process, the multilevel |BOM| is configured, and the top-level product has its
inventory counts, procurement method set, procurement lead times, and manufacturing operations
configured. From here, sales orders can include the top-level product, automatic procurement through
vendors or manufacturing can begin, the top-level product can be included in an **eCommerce** store.

.. tip::
   Continue configuring the database by setting up :doc:`sales quotations
   <../../../sales/sales/send_quotations/create_quotations>` and :doc:`eCommerce shop pages

>    <../../../websites/ecommerce/products/catalog>`

Reviewer 2: I think there may be an opportunity here to add an example of a product with everything configured (but a small, easy one!).  Something to just quickly show what you mean, maybe?

- **Option 2:** Activate the :guilabel:`Replenish on Order (MTO)` and :guilabel:`Manufacture` routes
  under the :guilabel:`Inventory` tab of the sublevel product's product form.

.. note::
   Learn more about :doc:`reordering rules <../../purchase/products/reordering>` and about
   :doc:`replenishing stock on order <../../inventory/warehouses_storage/replenishment/mto>`

Option 1 is more flexible than Option 2 and is recommended. Reordering rules do not directly link
demand to replenishment, allowing stock to be unreserved and reassigned as needed. The Replenish on
Order (MTO) route, however, uniquely links sublevel and top-level products, reserving quantities for
the confirmed top-level manufacturing order.

In both methods, sublevel products must be fully manufactured before starting the top-level product.

> Multilevel BoM setup flow

Reviewer 3: This section is SUCH an improvement! The way you’ve restructured it, along with the additional information, significantly enhances the reader’s understanding—not just of multi-level BoMs, but also how they connect to procurement and other parts of Odoo.

Amazing job! This update alone makes this PR fully deserving of a 5-point qualification. Your work has made this documentation far more relevant and impactful, truly giving it a new life. This is a fantastic example of a strong, meaningful overhaul! 🎉

=========================

In this flow, a multilevel |BOM| gets created, initial inventory is set, a 0/0/1 reordering rule is
created (the recommended production plan), lead times get set, and a production options get

> configured.

Reviewer 3: To reword this into active voice, how about something like:

The following section details how to set up multilevel BoMs, set the initial inventory, establish a :ref:`0/0/1 reordering rule <link to [0/0/1 section](https://runbot125.odoo.com/runbot/static/build/76203184-18-0/logs/build/html/applications/inventory_and_mrp/inventory/warehouses_storage/replenishment/reordering_rules.html#reordering-rule)> (the recommended production plan), configure lead times, and set up production options.

subassemblies, and the final product). To do this, open the **Inventory** app, and then find the
products with filters, the search bar, or scrolling, and then click on them to open their product
form. From here, click the :guilabel:`On Hand` smart button, :doc:`select the variant
<../../../sales/sales/products_prices/products/variants>` if that was configured, and then enter the
quantity on hand.

.. seealso::
   :doc:`../../inventory/warehouses_storage/inventory_management/count_products`

Configure the procurement method
--------------------------------

Now it is time to pick the procurement method this multilevel |BOM| uses. The two options below are
preferred, but unique circumstances may lead to another procurement method making more sense.

- **Option 1 (recommended):** Create *Reordering Rules* for the sublevel products and set both the
  minimum and maximum needed stock quantities to `0`.
- **Option 2:** Activate the :guilabel:`Replenish on Order (MTO)` and :guilabel:`Manufacture` routes
  under the :guilabel:`Inventory` tab of the sublevel product's product form.

.. tip::
   Learn more about :doc:`reordering rules <../../purchase/products/reordering>` and about

>    :doc:`replenishing stock on order <../../inventory/warehouses_storage/replenishment/mto>`

Reviewer 3: for consistency, let's change the admonition from tip to whatever admonition type you decide to go with on lines 86-88

Reordering rules are recommended because they do not tie the manufactured product to a specific
sales order, allowing that manufactured product to fulfill a different sales order if the original
gets cancelled.

Making the product to order is not recommended because the manufactured product cannot be used to
fulfill another sales order. However, this could be helpful if strict tracking is necessary for the
business.

Enter vendor and manufacturing lead times
-----------------------------------------

> Vendor and manufacturing lead times will lead to better predictions for production and order
> fulfillment. Set vendor lead times for components that are purchased. These can appear at any level

Reviewer 3: ```suggestion
Vendor and manufacturing lead times are used by Odoo to coordinate production and procurement actions to fulfill orders on time. Set vendor lead times for components that are purchased. These can appear at any level
```

in a multilevel |BOM| except the final product. Set manufacturing lead times for products that are
built using a |BOM|. These can appear at any level in a multilevel |BOM| except the lowest level
(when individual components are procured).

.. tip::
   Learn more about :doc:`vendor lead times and manufacturing lead times

>    <../../inventory/warehouses_storage/replenishment/lead_times>`, like how to set a lead time.

Reviewer 3: again for consistency, let's change the admonition from tip to whatever admonition type you decide to go with on lines 86-88

Build out operations to handle the production flow
--------------------------------------------------

First, determine the current manufacturing flow for the business, then match the corresponding Odoo
setup. The list below are only some of the configuration pieces that could be involved in this step.

- **Manufacturing Steps**: Consider how many manufacturing steps are taken (one-, two-, or
  three-step manufacturing).
- **Work Centers**: Decide if any :doc:`work centers <using_work_centers>` need to be configured.

> - **Master Production Schedules**: If manually planned manufacturing orders are needed (e.g., to

Reviewer 3: ```suggestion
- **Master Production Schedule**: If manually planned manufacturing orders are needed (e.g., to
```

  under the :guilabel:`Inventory` tab of the sublevel product's product form.

.. seealso::

>    `<../../inventory/warehouses_storage/replenishment/reordering_rules>`
>    `<../../inventory/warehouses_storage/replenishment/mto>`

Reviewer: Missing the role tag, and no need to include the doc path in `<...>` when not including link text, as the doc reference will use the referenced doc's H1 by default

<img width="677" alt="Screenshot 2025-03-14 at 11 34 28 AM" src="https://github.com/user-attachments/assets/d437269a-21e5-486d-82d2-15370b2de15f" />

```suggestion
   - :doc:`../../inventory/warehouses_storage/replenishment/reordering_rules`
   - :doc:`../../inventory/warehouses_storage/replenishment/mto`
```

Option 1 is more flexible than Option 2 and is recommended. Reordering rules do not directly link
demand to replenishment, allowing stock to be unreserved and reassigned as needed. The Replenish on
Order (MTO) route, however, uniquely links sublevel and top-level products, reserving quantities for
the confirmed top-level manufacturing order.

In both methods, sublevel products must be fully manufactured before starting the top-level product.

Multilevel BoM setup flow
=========================

The following section details how to set up multilevel BoMs, set the initial inventory, establish a

> :ref:`0/0/1 reordering rule <inventory/manufacturing/procurement-method>` (the recommended
> production plan), configure lead times, and set up production options.

Reviewer: Update ref to anchor
```suggestion
:ref:`0/0/1 reordering rule <manufacturing/advanced/procurement-method>` (the recommended production
plan), configure lead times, and set up production options.
```

Creating a 0/0/1 reordering rule for the sublevel products (minimum stock set at zero, maximum stock
set at zero, reorder one automatically) regardless of whether they are a component or subassembly is
the recommended approach for managing a multilevel |BOM|. This setup uses the **Inventory**,
**Manufacturing**, and **Purchase** apps.

.. important::
   This is only one example of how to set up a multilevel |BOM| in Odoo. Consider any unique
   circumstances that need to be addressed during the configuration, and make sure they are included
   in the setup. If any specific help is needed during setup, consider purchasing a `success pack
   <https://www.odoo.com/pricing-packs>`_.

Create the BoMs
---------------

> Follow the steps in the :ref:`Create a multilevel BoM <Create a Multilevel BoM>` section to build
> the |BOMs|.

Reviewer: Update ref to anchor
```suggestion
Follow the steps in the :ref:`Create a multilevel BoM <manufactuing/advanced/create-multilevel-bom>`
section to build the |BOMs|.
```

Make sure to build the multilevel |BOM| from the bottom up. Start by creating the lowest-level
component products in Odoo, then the subassembly products that those are used for, then the |BOM|
for that subassembly, and repeat until every level of the multilevel |BOM| is created.

Set the initial inventory
-------------------------

.. note::
   If there's no initial inventory to configure, then skip this section and begin configuring the
   procurement method for the multilevel |BOM|.

Update the quantity on hand for each of the products configured in the prior step (both components,
subassemblies, and the final product). To do this, open the **Inventory** app, and then find the
products with filters, the search bar, or scrolling, and then click on them to open their product
form. From here, click the :guilabel:`On Hand` smart button, :doc:`select the variant
<../../../sales/sales/products_prices/products/variants>` if that was configured, and then enter the
quantity on hand.

.. seealso::
   :doc:`../../inventory/warehouses_storage/inventory_management/count_products`

> .. _inventory/manufacturing/procurement-method:

Reviewer: This is a manufacturing app doc, not inventory :)
```suggestion
.. _manufacturing/advanced/procurement-method:
```

Configure the procurement method
--------------------------------

Now it is time to pick the procurement method this multilevel |BOM| uses. The two options below are
preferred, but unique circumstances may lead to another procurement method making more sense.

- **Option 1 (recommended):** Create *Reordering Rules* for the sublevel products and set both the
  minimum and maximum needed stock quantities to `0`.
- **Option 2:** Activate the :guilabel:`Replenish on Order (MTO)` and :guilabel:`Manufacture` routes
  under the :guilabel:`Inventory` tab of the sublevel product's product form.

.. seealso::

>    `<../../inventory/warehouses_storage/replenishment/reordering_rules>`
>    `<../../inventory/warehouses_storage/replenishment/mto>`

Reviewer: ```suggestion
   - :doc:`../../inventory/warehouses_storage/replenishment/reordering_rules`
   - :doc:`../../inventory/warehouses_storage/replenishment/mto`
```

Reordering rules are recommended because they do not tie the manufactured product to a specific
sales order, allowing that manufactured product to fulfill a different sales order if the original
gets cancelled.

Making the product to order is not recommended because the manufactured product cannot be used to
fulfill another sales order. However, this could be helpful if strict tracking is necessary for the
business.

Enter vendor and manufacturing lead times
-----------------------------------------

Vendor and manufacturing lead times are used by Odoo to coordinate production and procurement
actions to fulfill orders on time. Set vendor lead times for components that are purchased. These
can appear at any level in a multilevel |BOM| except the final product. Set manufacturing lead times
for products that are built using a |BOM|. These can appear at any level in a multilevel |BOM|
except the lowest level (when individual components are procured).

.. seealso::

>    `<../../inventory/warehouses_storage/replenishment/lead_times>`

Reviewer: ```suggestion
   :doc:`../../inventory/warehouses_storage/replenishment/lead_times`
```

setup. The list below are only some of the configuration pieces that could be involved in this step.

- **Manufacturing Steps**: Consider how many manufacturing steps are taken (one-, two-, or
  three-step manufacturing).
- **Work Centers**: Decide if any :doc:`work centers <using_work_centers>` need to be configured.
- **Master Production Schedule**: If manually planned manufacturing orders are needed (e.g., to
  handle seasonal demand), make a :doc:`master production schedule <../workflows/use_mps>` (MPS).

.. tip::
   Manufacturing operations is an art and a science, so configuring an established flow into Odoo is
   the recommended approach for this step. Read more about :doc:`manufacturing in Odoo
   <../../manufacturing>`

Configuration summary
---------------------

At the end of this process, the multilevel |BOM| is configured, and the top-level product has its
inventory counts, procurement method set, procurement lead times, and manufacturing operations
configured. From here, sales orders can include the top-level product, automatic procurement through
vendors or manufacturing can begin, the top-level product can be included in an **eCommerce** store.

.. seealso::

>    `<../../../sales/sales/sales_quotations/create_quotations>`
>    `<../../../websites/ecommerce/products/catalog>`

Reviewer: ```suggestion
   - :doc:`../../../sales/sales/sales_quotations/create_quotations`
   - :doc:`../../../websites/ecommerce/products/catalog`
```
